librad/src/id.rs

```diff
@@ -214,8 +284,109 @@ pub trait Entity {
             Err(Error::SignatureVerificationFailed)
         }
     }
+
+    fn check_validity(&self, resolver: &impl Resolver<User>) -> Result<(), Error> {
+        let mut keys = HashSet::<PublicKey>::from_iter(self.keys());
+        let mut users = HashSet::<RadicleUri>::from_iter(self.certifiers());
+
+        for (k, s) in self.signatures() {
+            self.check_signature(k, &s.by, &s.sig, resolver)?;
+            match &s.by {
+                Signatory::OwnedKey => {
+                    keys.remove(k);
+                },
+                Signatory::User(user) => {
+                    users.remove(&user);
+                },
+            }
+        }
+        if keys.len() > 0 || users.len() > 0 {
+            Err(Error::SignatureMissing)
+        } else {
+            Ok(())
+        }
+    }
+
+    fn is_valid(&self, resolver: &impl Resolver<User>) -> bool {
+        self.check_validity(resolver).is_ok()
+    }
+
+    fn check_update(&self, previous: &Self) -> Result<(), UpdateVerificationError> {
+        if self.revision() <= previous.revision() {
+            return Err(UpdateVerificationError::NonMonotonicRevision);
+        }
+
+        let retained_keys = self.keys().filter(|k| previous.has_key(k)).count();
+        let total_keys = self.keys_count();
+        let added_keys = total_keys - retained_keys;
+        let removed_keys = previous.keys_count() - retained_keys;
+        let quorum_keys = total_keys / 2;
+
+        if added_keys > quorum_keys {
+            return Err(UpdateVerificationError::NoCurrentQuorum);
+        } else if removed_keys > quorum_keys {
+            return Err(UpdateVerificationError::NoPreviousQuorum);
+        }
+
+        let retained_certifiers = self
+            .certifiers()
+            .filter(|c| previous.has_certifier(c))
+            .count();
+        let total_certifiers = self.certifiers_count();
+        let added_certifiers = total_certifiers - retained_certifiers;
+        let removed_certifiers = previous.certifiers_count() - retained_certifiers;
+        let quorum_certifiers = total_certifiers / 2;
+
+        if added_certifiers > quorum_certifiers {
+            return Err(UpdateVerificationError::NoCurrentQuorum);
+        } else if removed_certifiers > quorum_certifiers {
+            return Err(UpdateVerificationError::NoPreviousQuorum);
+        }
+
+        Ok(())
+    }
+
+    fn check_history<E, I, II>(
+        uri: &RadicleUri,
+        resolver: &impl Resolver<User>,
+        revisions_resolver: &impl RevisionsResolver<E, I, II>,
+    ) -> Result<(), HistoryVerificationError>
+    where
+        E: Entity,
+        I: Iterator<Item = E>,
+        II: IntoIterator<Item = E, IntoIter = I> + Sized,
+    {
+        let mut revisions = revisions_resolver.resolve_revisions(uri).into_iter();
+
+        let current = revisions.next();
+        let mut current = match current {
```

Comment:
```suggestion
        let mut current = match revisions.next() {
```

Could the shadowing here by doing the above